components/usage/pkg/db/cost_center.go

```diff
@@ -27,36 +29,136 @@ type CostCenter struct {
+	NextBillingTime VarcharTime     `gorm:"column:nextBillingTime;type:varchar;size:255;" json:"nextBillingTime"`
+	LastModified    time.Time       `gorm:"->:column:_lastModified;type:timestamp;default:CURRENT_TIMESTAMP(6);" json:"_lastModified"`
 }
 
 // TableName sets the insert table name for this struct type
 func (d *CostCenter) TableName() string {
 	return "d_b_cost_center"
 }
 
-func GetCostCenter(ctx context.Context, conn *gorm.DB, attributionId AttributionID) (*CostCenter, error) {
+type DefaultSpendingLimit struct {
+	ForTeams int32 `json:"forTeams"`
+	ForUsers int32 `json:"forUsers"`
+}
+
+func NewCostCenterManager(conn *gorm.DB, cfg DefaultSpendingLimit) *CostCenterManager {
+	return &CostCenterManager{
+		conn: conn,
+		cfg:  cfg,
+	}
+}
+
+type CostCenterManager struct {
+	conn *gorm.DB
+	cfg  DefaultSpendingLimit
+}
+
+// GetOrCreateCostCenter returns the latest version of cost center for the given attributionID.
+// This method creates a codt center and stores it in the DB if there is no preexisting one.
+func (c *CostCenterManager) GetOrCreateCostCenter(ctx context.Context, attributionID AttributionID) (CostCenter, error) {
+	logger := log.WithField("attributionId", attributionID)
+
+	result, err := getCostCenter(ctx, c.conn, attributionID)
+	if err != nil {
+		if errors.Is(err, CostCenterNotFound) {
+			logger.Info("No existing cost center. Creating one.")
+			defaultSpendingLimit := c.cfg.ForUsers
+			if !attributionID.IsUser() {
+				defaultSpendingLimit = c.cfg.ForTeams
+			}
+			result = CostCenter{
+				ID:              attributionID,
+				CreationTime:    NewVarcharTime(time.Now()),
+				BillingStrategy: CostCenter_Other,
+				SpendingLimit:   defaultSpendingLimit,
+				NextBillingTime: NewVarcharTime(time.Now().AddDate(0, 1, 0)),
+			}
+			err := c.conn.Save(&result).Error
+			if err != nil {
+				return CostCenter{}, err
+			}
+		} else {
+			return CostCenter{}, err
+		}
+	}
+
+	return result, nil
+}
+
+func getCostCenter(ctx context.Context, conn *gorm.DB, attributionId AttributionID) (CostCenter, error) {
 	db := conn.WithContext(ctx)
 
 	var results []CostCenter
 	db = db.Where("id = ?", attributionId).Order("creationTime DESC").Limit(1).Find(&results)
 	if db.Error != nil {
-		return nil, fmt.Errorf("failed to get cost center: %w", db.Error)
+		return CostCenter{}, fmt.Errorf("failed to get cost center: %w", db.Error)
 	}
 	if len(results) == 0 {
-		return nil, CostCenterNotFound
+		return CostCenter{}, CostCenterNotFound
 	}
 	costCenter := results[0]
-	return &costCenter, nil
+	return costCenter, nil
 }
 
-func SaveCostCenter(ctx context.Context, conn *gorm.DB, costCenter *CostCenter) (*CostCenter, error) {
-	db := conn.WithContext(ctx)
-	costCenter.CreationTime = NewVarcharTime(time.Now())
-	db = db.Save(costCenter)
+func (c *CostCenterManager) SaveCostCenter(ctx context.Context, costCenter CostCenter) (CostCenter, error) {
```

Comment:
From a consumer of the `CostCenterManager` API perspective, it's quite unclear what the difference is between `GetOrCreate` and `SaveCostCenter`. Why would I use one over the other?

I think here, we can rename it to
```suggestion
func (c *CostCenterManager) UpdateCostCenter(ctx context.Context, costCenter CostCenter) (CostCenter, error) {
```
To better communicate that this is intended to work on an existing record(even if it needs to be created under the hood)

Reply:
I aligned with the naming of gorm but have no problem changing this to 'Update..'.